src/topology/uniform_space/uniform_embedding.lean

```diff
@@ -1,29 +1,56 @@
 /-
 Copyright (c) 2017 Johannes Hölzl. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
-Authors: Johannes Hölzl, Sébastien Gouëzel
+Authors: Johannes Hölzl, Sébastien Gouëzel, Patrick Massot
 
 Uniform embeddings of uniform spaces. Extension of uniform continuous functions.
 -/
-import topology.uniform_space.cauchy
+import topology.uniform_space.cauchy topology.uniform_space.separation
 
 open filter topological_space lattice set classical
 local attribute [instance, priority 0] prop_decidable
-variables {α : Type*} {β : Type*} {γ : Type*} [uniform_space α]
+local notation `𝓤` := uniformity
+
+section
+variables {α : Type*} {β : Type*} {γ : Type*}
+          [uniform_space α] [uniform_space β] [uniform_space γ]
 universe u
 
-local notation `𝓤` := uniformity
+structure uniform_inducing (f : α → β) : Prop :=
+(comap_uniformity : comap (λx:α×α, (f x.1, f x.2)) (𝓤 β) = 𝓤 α)
+
+def uniform_inducing.mk' {f : α → β} (h : ∀ s, s ∈ 𝓤 α ↔
+    ∃ t ∈ 𝓤 β, ∀ x y : α, (f x, f y) ∈ t → (x, y) ∈ s) : uniform_inducing f :=
+⟨by simp [eq_comm, filter.ext_iff, subset_def, h]⟩
 
-def uniform_embedding [uniform_space β] (f : α → β) :=
-function.injective f ∧
-comap (λx:α×α, (f x.1, f x.2)) (𝓤 β) = 𝓤 α
+lemma uniform_inducing.comp {f : α → β} (hf : uniform_inducing f)
+  {g : β → γ} (hg : uniform_inducing g) : uniform_inducing (g ∘ f) :=
+⟨ by rw [show (λ (x : α × α), ((g ∘ f) x.1, (g ∘ f) x.2)) =
+         (λ y : β × β, (g y.1, g y.2)) ∘ (λ x : α × α, (f x.1, f x.2)), by ext ; simp,
+        ← filter.comap_comap_comp, hg.1, hf.1]⟩
 
-theorem uniform_embedding_def [uniform_space β] {f : α → β} :
+structure uniform_embedding (f : α → β) extends uniform_inducing f : Prop :=
+(inj : function.injective f)
+
+lemma uniform_embedding.comp {f : α → β} (hf : uniform_embedding f)
+  {g : β → γ} (hg : uniform_embedding g) : uniform_embedding (g ∘ f) :=
+{ inj := function.injective_comp hg.inj hf.inj,
+  ..hf.to_uniform_inducing.comp hg.to_uniform_inducing }
+
+theorem uniform_embedding_def {f : α → β} :
   uniform_embedding f ↔ function.injective f ∧ ∀ s, s ∈ 𝓤 α ↔
     ∃ t ∈ 𝓤 β, ∀ x y : α, (f x, f y) ∈ t → (x, y) ∈ s :=
-by rw [uniform_embedding, eq_comm, filter.ext_iff]; simp [subset_def]
+begin
+  split, rintro ⟨⟨h⟩, h'⟩,
```

Comment:
I did not open the proof in Lean, but a proof starting with split and without braces looks suspicious.